nncf/experimental/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,695 @@
+        self.op_addresses = op_addresses
+        self.modules = modules
+        self.block_type = block_type
+
+
+class SearchGraph:
+    """
+    A wrapper over the graph, which represents the DNN execution graph transformed
+    by pattern matching, merging nodes and inserting auxiliary nodes.
+
+    """
+    ACTIVATION_INPUT_SHAPE_ATTR = 'activation_input_shape'
+    ACTIVATION_OUTPUT_SHAPE_ATTR = 'activation_output_shape'
+    IS_MERGED_NODE_ATTR = 'is_merged'
+    IS_DUMMY_NODE_ATTR = 'is_dummy'
+    KEY_NODE_ATTR = 'key'
+    MERGED_NODES_NODE_ATTR = 'merged_nodes'
+    TYPE_NODE_ATTR = 'type'
+    DUMMY_POSTFIX = " dummy"
+
+
+    def __init__(self, nx_merged_graph: nx.DiGraph):
+        self._nx_graph = nx_merged_graph
+        old_merged_graph_nodes = deepcopy(self._nx_graph._node)
+        for node_key in old_merged_graph_nodes:
+            # pylint: disable=protected-access
+            next_nodes = self._nx_graph._succ[node_key]
+            if len(list(next_nodes)) > 1:
+                self._insert_dummy_node(node_key)
+
+    def _nx_node_to_sgraph_node(self, nx_node_key: str, nx_node_attrs: Dict):
+        return SearchGraphNode(nx_node_key, nx_node_attrs)
+
+    def get_node_by_key(self, node_key: str) -> SearchGraphNode:
+        """
+        :param node_key: key (node_name) of the node.
+        :return: SearchGraphNode in a graph with such key.
+        """
+        return SearchGraphNode(node_key, self._nx_graph.nodes[node_key])
+
+    def get_all_nodes(self) -> List[SearchGraphNode]:
+        """
+        Returns list of all graph nodes.
+        """
+        all_nodes = []
+        for node_key, node_attrs in self._nx_graph.nodes.items():
+            all_nodes.append(SearchGraphNode(node_key, node_attrs))
+        return all_nodes
+
+    def get_next_nodes(self, node_key: str) -> List[SearchGraphNode]:
+        """
+        Returns consumer nodes of provided node key.
+
+        :param node_key: Producer node key.
+        :return: List of consumer nodes of provided node.
+        """
+        next_node_keys = self._nx_graph.succ[node_key]
+        return [self.get_node_by_key(node_key) for node_key in next_node_keys]
+
+    def set_node_attr(self, node_key: str, name_attr: str, value_attr: str):
+        """
+        Set value of attribute by name for a given node with the same key.
+        """
+        self._nx_graph.nodes[node_key][name_attr] = value_attr
+
+    def get_prev_nodes(self, node_key: str) -> List[SearchGraphNode]:
+        """
+        Returns producer nodes of provided node key.
+
+        :param node_key: Consumer node key.
+        :return: List of producers nodes of provided node.
+        """
+        prev_node_keys = self._nx_graph.pred[node_key]
+        return [self.get_node_by_key(node_key) for node_key in prev_node_keys]
+
+    def get_prev_edges(self, node_key: str) -> Dict:
+        return self._nx_graph.pred[node_key]
+
+    def get_next_edges(self, node_key: str) -> Dict:
```

Comment:
Need to clarify the key and value types here and in other places with `Dict` or `List` or `Set` as a type hint

Reply:
Done.